src/lib/cost_model/abstract_cost_estimator.cpp

```diff
@@ -26,11 +30,48 @@ Cost AbstractCostEstimator::estimate_plan_cost(const std::shared_ptr<AbstractLQP
       continue;
     }
 
-    cost += _estimate_node_cost(current_node);
+    // Look up this subplan in cache, if a cache is present
+    if (cost_estimation_cache) {
+      const auto cost_estimation_cache_iter = cost_estimation_cache->find(current_node);
+      if (cost_estimation_cache_iter != cost_estimation_cache->end()) {
+        auto subplan_already_visited = false;
+
+        auto subplan_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>{};
+
+        // Check whether the cache entry can be used: This is only the case if the entire subplan has not yet been
+        // visited. If it already has been visited, and we'd use the cache entry anyway, costs would get counted twice
```

Comment:
Changed to 

```
    /**
     * Look up this subplan in cache, if a cache is present.
     * Needs to pay special attention to diamond shapes in the LQP to avoid costing nodes multiple times.
     */
```